src/app/GitUI/Editor/Diff/DiffHighlightService.cs

```diff
@@ -126,115 +127,137 @@ public override void AddTextHighlighting(IDocument document)
-            && DoesLineStartWith(document, endLine.Offset, prefixStr, invertMatch);
-            line++)
+        foreach (ISegment segment in GetAllLines(DiffLineType.Header))
         {
-            endLine = document.GetLineSegment(line);
+            document.MarkerStrategy.AddMarker(CreateTextMarker(segment.Offset, segment.Length, AppColor.DiffSection.GetThemeColor()));
         }
+    }
 
-        line = Math.Max(0, line - 2);
-        endLine = document.GetLineSegment(line);
+    /// <summary>
+    ///  Matches related removed and added lines in a consecutive block of a patch document and marks identical parts dimmed.
+    /// </summary>
+    private void MarkInlineDifferences(IDocument document)
+    {
+        int index = 0;
+        List<DiffLineInfo> diffLines = [.. _diffLinesInfo.DiffLines.Values.OrderBy(i => i.LineNumInDiff)];
+        bool found = false;
 
-        document.MarkerStrategy.AddMarker(new TextMarker(lineSegment.Offset,
-            (endLine.Offset + endLine.TotalLength) -
-            lineSegment.Offset, TextMarkerType.SolidBlock, color,
-            ColorHelper.GetForeColorForBackColor(color)));
+        const int diffContentOffset = 1; // in order to skip the prefixes '-' / '+'
+        MarkerStrategy markerStrategy = document.MarkerStrategy;
 
-        return;
+        // Process the next blocks of removed / added diffLines and mark in-line differences
+        while (index < diffLines.Count)
+        {
+            found = false;
 
-        bool DoesLineStartWith(IDocument document, int offset, string prefixStr, bool invertMatch)
-            => invertMatch ^ LinePrefixHelper.DoesLineStartWith(document, offset, prefixStr);
-    }
+            // git-diff presents the removed lines directly followed by the added in a "block"
+            IReadOnlyList<ISegment> linesRemoved = GetBlockOfLines(diffLines, DiffLineType.Minus, ref index, ref found);
+            IReadOnlyList<ISegment> linesAdded = GetBlockOfLines(diffLines, DiffLineType.Plus, ref index, ref found);
 
-    private void SetText(ref string text)
-    {
-        if (!_useGitColoring)
-        {
-            return;
+            foreach (TextMarker marker in GetDifferenceMarkers(GetText, linesRemoved, linesAdded, diffContentOffset))
+            {
+                markerStrategy.AddMarker(marker);
+            }
         }
 
-        StringBuilder sb = new(text.Length);
-        AnsiEscapeUtilities.ParseEscape(text, sb, _textMarkers);
+        return;
 
-        text = sb.ToString();
+        string GetText(ISegment line)
+        {
+            int len = line.Length - diffContentOffset;
+            if (line.Offset + line.Length >= document.TextLength)
+            {
+                // This is likely a test, where last line do not have a newline
+                --len;
+            }
+
+            return document.GetText(line.Offset + diffContentOffset, len);
+        }
     }
 
+    private List<ISegment> GetAllLines(DiffLineType diffLineType)
```

Comment:
It is `private` and iterated once only, i.e. no need for `.ToList()`.
```suggestion
    private IEnumerable[] GetAllLines(DiffLineType diffLineType)
```